envoy.dependency.check/envoy/dependency/check/abstract/dependency.py

```diff
@@ -0,0 +1,179 @@
+
+    def __lt__(self, other: "ADependency") -> bool:
+        return self.id < other.id
+
+    def __str__(self):
+        return f"{self.id}@{self.version}"
+
+    @async_property(cache=True)
+    async def commits_since_current(self) -> int:
+        """Commits since current commit/tag."""
+        count = await self.repo.commits(
+            since=await self.release.timestamp_commit).total_count
+        return count and count - 1 or count
+
+    @cached_property
+    def cpe(self) -> Optional[str]:
+        """Configured CPE for this dependency."""
+        return (
+            str(self.metadata["cpe"])
+            if self.metadata.get("cpe", "N/A") != "N/A"
+            else None)
+
+    @cached_property
+    def github_url(self) -> Optional[str]:
+        """Github URL."""
+        for url in self.urls:
+            if url.startswith('https://github.com/'):
+                return url
+
+    @cached_property
+    def github_version(self) -> str:
+        """Github version, as parsed from the URL."""
+        if self.url_components[5] != 'archive':
+            # Release tag is a path component.
+            # assert (components[5] == 'releases')
+            return self.url_components[7]
+        # Only support .tar.gz, .zip today. Figure out the release tag
+        # from this filename.
+        if self.url_components[-1].endswith('.tar.gz'):
+            return self.url_components[-1][:-len('.tar.gz')]
+        # assert (components[-1].endswith('.zip'))
+        return self.url_components[-1][:-len('.zip')]
+
+    @property
+    def github_version_name(self) -> str:
+        """Github version, truncated to 7 char if its sha_hash."""
+        return (
+            self.github_version[0:7]
+            if not self.release.tagged
+            else self.github_version)
+
+    @async_property
+    async def has_recent_commits(self) -> bool:
+        """Flag indicating whether there are more recent commits than the
+        current pinned commit."""
+        return await self.recent_commits > 1
+
+    @async_property(cache=True)
+    async def newer_release(
+            self) -> Optional["abstract.ADependencyGithubRelease"]:
+        """Release with highest semantic version if newer than the current
+        release, or where pin is to tag or commit."""
+        # TODO: consider adding `newer_tags` for deps that only create
+        #   tags and not releases (eg tclap)
+        newer_release = await self.repo.highest_release(
+            since=await self.release.timestamp)
+        return (
+            self.release_class(
+                self.repo,
+                newer_release.tag_name,
+                release=newer_release)
+            if (newer_release
+                and (version.parse(newer_release.tag_name)
+                     != self.release.version))
+            else None)
+
+    @property
+    def organization(self) -> str:
+        """Github organization name."""
+        return self.url_components[3]
```

Comment:
should we have all these indices defined as constants so that it is easier to understand?

Reply:
this is one of the bits i least like about this implementation 

what i would like to do is add a `GithubURLParser` that can tell you

- the type of url
- org/user/repo
- additional info according to type

and that would raise appropriate errors

i havent found anything that does this already, but i havent looked that hard either

as this will require a bit of thought - for now i would prefer to keep the existing implementation (im just handling a bit of error handling to it)

if its a blocker tho, we can rethink the url parsing

one consideration here is that the URL is the SoA for the repo connect - this means that configs like here https://github.com/envoyproxy/envoy/pull/19473/files#diff-a2ca502474bb8ec5f092176f7ee6d2f11fc395ebb00e1717d6fbad9b16ee8fd8R481 wont work 

(i will comment this shortly 8/)

tldr i think this needs quite a bit of a rethink - but this is close to what it does now and works to our current expectations, but i would prefer not to formalize it